direct_connect.go

```diff
@@ -0,0 +1,187 @@
+package main
```

Comment:
I didn't add tests for this yet, but I think the best/correct thing to do is an integration test, because there's no good way to do some unit test for it.